In [42]:
# Machine learning phase: training a model to predict passenger survival
# Notebook uses the cleaned and feature-engineered dataset created in the previous step
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
# Ladoing the dataset that contains all engineered features
df = pd.read_csv('Titanic_for_ML.csv')
print('--Data load--')

--Data load--


In [43]:
# Verify the dataset shape
print(f'Rows: {df.shape[0]} -- Columns: {df.shape[1]}')

Rows: 891 -- Columns: 26


In [44]:
# Inspection of the ml data frame head and first rows
df.head()

,Survived,Pclass,SibSp,Parch,FamilySize,isAlone,Sex_female,Sex_male,Embarked_C,Embarked_Q,Embarked_S,hasCabin,Title_Master,Title_Miss,Title_Mr,Title_Mrs,Title_Rare,FareCategory_Cheap,FareCategory_Medium,FareCategory_Expensive,FareCategory_Luxury,AgeGroup_Child,AgeGroup_Teen,AgeGroup_Young Adult,AgeGroup_Adult,AgeGroup_Senior
0,0,3,1,0,2,0,0,1,0,0,1,0,0,0,1,0,0,1,0,0,0,0,0,1,0,0
1,1,1,1,0,2,0,1,0,1,0,0,1,0,0,0,1,0,0,0,0,1,0,0,0,1,0
2,1,3,0,0,1,1,1,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0
3,1,1,1,0,2,0,1,0,0,0,1,1,0,0,0,1,0,0,0,0,1,0,0,1,0,0
4,0,3,0,0,1,1,0,1,0,0,1,0,0,0,1,0,0,0,1,0,0,0,0,1,0,0


In [45]:
# Splitting into X the features and y the target
# Features - all relevant input variables
X = df.drop(columns='Survived')
# Target - extracting survived as target variable
y = df['Survived']

In [46]:
# Splitting data into 80% train and 20% validation set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Checking the size of train and test set with shape function
print('--DATA CHECK--')
print(f'Number of Passangers in the train set: {X_train.shape[0]}')
print(f'Columns in the train set: {X_train.shape[1]}')
print(f'Number of Passangers in the test set: {y_test.shape[0]}')

--DATA CHECK--
Number of Passangers in the train set: 712
Columns in the train set: 25
Number of Passangers in the test set: 179


In [47]:
# Creating the first model - LogisticRegression with iter of 1000
model = LogisticRegression(max_iter=1000)
# Training the model on the training data
model.fit (X_train, y_train)
print('--done--')

--done--


In [48]:
# Using trained model to make predictions
prediction = model.predict(X_test)
#model evaluations metrics
accuracy = accuracy_score(y_test, prediction)
# Checking accuracy score
print(f'Accuracy Score: {accuracy * 100:.2f}%')

Accuracy Score: 82.68


In [49]:
# Checking the coefficients
coefficients = pd.DataFrame({'Feature': X.columns, 'Points': model.coef_[0]})
# Sort the coefficients by points
coefficients = coefficients.sort_values(by='Points', ascending=False)
print(coefficients)

                   Feature    Points
11            Title_Master  1.423753
5               Sex_female  0.870789
20          AgeGroup_Child  0.788799
14               Title_Mrs  0.777731
10                hasCabin  0.609153
19     FareCategory_Luxury  0.276382
7               Embarked_C  0.244471
22    AgeGroup_Young Adult  0.077574
18  FareCategory_Expensive  0.056471
8               Embarked_Q  0.020750
2                    Parch -0.097946
21           AgeGroup_Teen -0.110734
17     FareCategory_Medium -0.122467
16      FareCategory_Cheap -0.201384
12              Title_Miss -0.205214
4                  isAlone -0.251702
1                    SibSp -0.253677
9               Embarked_S -0.256219
24         AgeGroup_Senior -0.308969
3               FamilySize -0.342622
23          AgeGroup_Adult -0.437668
0                   Pclass -0.627226
15              Title_Rare -0.731684
6                 Sex_male -0.861788
13                Title_Mr -1.255584


**Note:** The train above was with the LogisticRegression Model, additionally in the following code the RandomForestClassifier will be importet

In [50]:
# Import RandomForestClassifier
from sklearn.ensemble import RandomForestClassifier

# Implement the RandomForest model and set random_state to 42
rf_model = RandomForestClassifier(random_state=42)
# Training the model in the training data
rf_model.fit(X_train, y_train)
print('done')

done


In [51]:
# Using trained model to make predictions
rf_prediction = rf_model.predict(X_test)
# Ealuation model performance
rf_accuracy = accuracy_score(y_test, rf_prediction)
# Checking the accuracy
print(f'Accuracy Score: {rf_accuracy*100:.2f}%')

Accuracy Score: 82.12%


In [52]:
# Checking the feature importance from RandomForestClassifier
rf_importance = pd.DataFrame({'Feature': X.columns, 'Rating': rf_model.feature_importances_})
# Sorting the importance by Rating
rf_importance = rf_importance.sort_values(by='Rating', ascending=False)
print(rf_importance)

                   Feature    Rating
6                 Sex_male  0.140113
0                   Pclass  0.102373
5               Sex_female  0.101266
13                Title_Mr  0.094842
3               FamilySize  0.075812
10                hasCabin  0.060326
1                    SibSp  0.048844
14               Title_Mrs  0.037880
2                    Parch  0.035143
12              Title_Miss  0.030456
19     FareCategory_Luxury  0.027566
9               Embarked_S  0.024622
17     FareCategory_Medium  0.023811
20          AgeGroup_Child  0.023660
22    AgeGroup_Young Adult  0.023453
18  FareCategory_Expensive  0.022864
7               Embarked_C  0.022393
16      FareCategory_Cheap  0.020643
23          AgeGroup_Adult  0.018153
21           AgeGroup_Teen  0.015475
4                  isAlone  0.014008
11            Title_Master  0.012486
8               Embarked_Q  0.011521
15              Title_Rare  0.006630
24         AgeGroup_Senior  0.005660


**Note:** Used RandomForestClassifier with the train and test data and got and Accurancy Score of 82.12% which is a little less than with the LogisticRegression Model. Importing KNeighborsClassifier to check with this model

In [53]:
# Import KNeighborsClassfier 
from sklearn.neighbors import KNeighborsClassifier

# Implement the knn model
knn_model = KNeighborsClassifier(n_neighbors=5)
# Train the model on the training data
knn_model.fit(X_train, y_train)
print('done')

done


In [55]:
# Predict survival rate on the validation set
knn_predict = knn_model.predict(X_test)
# Checking accuracy on the validation set
knn_accur = accuracy_score(y_test, knn_predict)
print(f'Score: {knn_accur*100:.2f}%')

Score: 81.56%


**Note:** The KNeighborsClassifier model reached an accuracy score of 81.56% used with 5 n_neighbors. The model has the most less accuracy score from all three used models